### PR TITLE
feat: introduce boot plugin concept

### PR DESCRIPTION
**Pull request by jhaeyaert**
**Issue**

https://gravitee.atlassian.net/browse/ARCHI-297

**Description**

This PR introduces the concept of a boot plugin that can be loaded at the beginning of the startup phase. This allows for some particular plugins such as secret provider plugins to be loaded and fully set up before initializing the rest of the application (e.g.: application Spring beans that rely on secret resolution).

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `3.0.0-archi-297-boot-plugin-loading-isolation-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/plugin/gravitee-plugin/3.0.0-archi-297-boot-plugin-loading-isolation-SNAPSHOT/gravitee-plugin-3.0.0-archi-297-boot-plugin-loading-isolation-SNAPSHOT.zip)

### CODE REVIEW

**Review by remibaptistegio — CHANGES_REQUESTED** [review]

File: gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/AbstractPluginEventListener.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            .filter(pluginHandler -> pluginHandler.canHandle(plugin))
+            .forEach(pluginHandler -> {
+                log.debug("Plugin {} has been managed by {}", plugin.id(), pluginHandler.getClass());
+                pluginHandler.handle(plugin);
+            });
+
+        deployedPlugins.add(plugin);
+    }
+
+    private static class PluginKey {
```
[the remark this answers is not in the document]
Comment (jhaeyaert) [reply gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/AbstractPluginEventListener.java:129]: Yep, I didn't refactor this part while I was extracting it into an abstract class. Well spotted 👍

File: gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/PluginRegistryImpl.java [hunk 2]
```diff
@@ -77,36 +77,59 @@ public class PluginRegistryImpl extends AbstractService<PluginRegistry> implemen
...
         super.doStart();
 
-        if (!init) {
-            log.info("Initializing plugin registry.");
-            this.init();
-            log.info("Plugins have been loaded and installed.");
-        } else {
-            log.warn("Plugin registry has already been initialized.");
-        }
+        init();
```
[the remark this answers is not in the document]
Comment (jhaeyaert) [reply gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/PluginRegistryImpl.java:115]: I wanted to keep the bootstrap optional so we can still use the registry as we did previously and only call `start()` on it. There is a flag that prevents the init() to be executed twice. Do you see any issue if I let it like this?

[the remark this answers is not in the document]
Comment (remibaptistegio) [reply gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/PluginRegistryImpl.java:115]: Ok I see, I thought bootstrap will always be called.

File: gravitee-plugin-core/src/test/java/io/gravitee/plugin/core/internal/PluginRegistryTest.java [hunk 3]
```diff
@@ -256,12 +281,59 @@ void should_return_empty_when_no_plugins_for_type() throws Exception {
...
+        PluginRegistryImpl pluginRegistry = initPluginRegistry("/io/gravitee/plugin/with-dependencies/");
+
+        pluginRegistry.bootstrap();
+
+        verify(eventManager, times(pluginRegistry.plugins().size())).publishEvent(eq(PluginEvent.BOOT_DEPLOYED), any());
+        verify(eventManager).publishEvent(eq(PluginEvent.BOOT_ENDED), any());
+    }
+
+    @Test
+    void should_init_once_when_bootstrap_then_start() throws Exception {
```
[the remark this answers is not in the document]
Comment (jhaeyaert) [reply gravitee-plugin-core/src/test/java/io/gravitee/plugin/core/internal/PluginRegistryTest.java:316]: I renamed it `should_emit_boot_ended_event_only_once_when_bootstrap_and_then_start`, does it make it clearer?

**Review by remibaptistegio — APPROVED** [review]

**Pull request merged**